### ISSUE TITLE
Update the source links in GA4 Reporting widgets to use the correct GA4 report URLs.

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#6639]
## Feature Description

As discussed on Slack, we need to determine and use the correct URLs for the source links for the GA4 Reporting widgets.

In order not to block development, we can initially use the UA version of the URLs, and then update them once we have specced the GA4 URLs.

This outcome of this issue should be to have updated the various GA4 Reporting widgets with the correct source link URLs. Further detail can be added once the URLs are known, this is currently with @marrrmarrr to spec them out.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The "deep links" for source URLs on the **Main Dashboard** should be updated for GA4-sourced metrics/widgets to the following:
  * **All Traffic Widget**:
    * If the "Channels" tab is selected, update to [Reports > Life cycle > Acquisition > Traffic](https://support.google.com/analytics/answer/12923437?hl=en&ref_topic=9303476), eg: `https://analytics.google.com/analytics/web/#/p${propertyId}/reports/dashboard?params=_u..nav%3Dmaui&r=lifecycle-acquisition-overview&ruid=lifecycle-acquisition-overview,life-cycle,acquisition&collectionId=life-cycle`
    * If the "Countries" tab is selected, update to [Reports > Demographics > Overview > Users by country](https://support.google.com/analytics/answer/9353692), eg: `https://analytics.google.com/analytics/web/#/p${propertyId}/reports/dashboard?params=_u..nav%3Dmaui&r=user-demographics-overview&ruid=user-demographics-overview,user,demographics&collectionId=user`
    * If the  "Devices" tab is selected, update to [Reports > Tech > Tech details > Device category](https://support.google.com/analytics/answer/9353571), eg: `https://analytics.google.com/analytics/web/#/p${propertyId}/reports/explorer?params=_u..nav%3Dmaui%26_r.explorerCard..seldim%3D%5B%22deviceCategory%22%5D&r=user-technology-detail&ruid=user-technology-detail,user,technology&collectionId=user`
  * **Search Traffic Widget**
      * The first two metrics ("Total Impressions" and "Total Clicks") currently point to Search Console, so no changes should be made.
      * For the "Total Users from Search" metric, update to Reports > Life cycle > Acquisition > Traffic Acquisition report, with an added filter for `sessionSource` as `google` (example link), eg: `https://analytics.google.com/analytics/web/#/p${propertyId}/reports/explorer?params=_u..nav%3Dmaui%26_r..dataFilters%3D%5B%7B%22type%22:1,%22fieldName%22:%22sessionSource%22,%22evaluationType%22:3,%22expressionList%22:%5B%22google%22%5D,%22complement%22:false,%22isCaseSensitive%22:false,%22expression%22:%22%22%7D%5D&r=lifecycle-traffic-acquisition-v2&ruid=lifecycle-traffic-acquisition-v2,life-cycle,acquisition&collectionId=life-cycle`
  * **Top Search Queries Widget**
    * This currently points to Search Console, so no changes should be made.
  * **Top Content Widget**
    * Update to [Reports > Life cycle > Engagement > Pages and Screens](https://support.google.com/analytics/answer/12926732?hl=en) report, eg. `https://analytics.google.com/analytics/web/#/p${propertyId}/reports/explorer?params=_u..nav%3Dmaui&r=all-pages-and-screens&ruid=all-pages-and-screens,life-cycle,engagement&collectionId=life-cycle`
  * **Top Earning Pages Widget**
    * No change should be made as this requires GA4 and AdSense be integrated.
* Additionally, "deep links" for source URLs on the **Entity Dashboard** should be updated for GA4-sourced metrics/widgets to the following:
  * **All Traffic Widget** updates should mirror the Main Dashboard updates above, but include the `unifiedPagePathScreen` metric to show only data for the current page (that matches the page the user is viewing on the Entity Dashboard)
  * **Overall page metrics** should update to [Reports > Life cycle > Engagement > Pages and screens](https://support.google.com/analytics/answer/12926732?hl=en) the same as the "Top Content" metric above, but include the `unifiedPagePathScreen` filter for the current page.
  * Other metrics are already covered by the Main Dashboard or aren't needed.

See https://docs.google.com/document/d/17RKIrdm5PBK8j6rP0GV_2QlztSLXqYQX8ikz3kQmrAo/edit?pli=1&resourcekey=0-njOU7bn1dhvi5Yp2jeX5zw#bookmark=id.9e5lpgtwjov3 for metrics to use to replace these links.

## Implementation Brief

* Create `assets/js/modules/analytics-4/datastore/service.js`:
    * This will hold the `getServiceURL` and `getServiceReportURL` selectors for Analytics 4 module. 
    * See the aforementioned selectors of Analytics module for inspiration and create the selectors for GA4.
    * They should be structured in a way to return the correct deep links mentioned in the AC. 
        * the `propertyID` should be replaced with GA4 property id. 
        * Notice that some of the query parameters are JSON encoded JS objects that are then URL encoded. We should construct them similarly. Thanks @techanvil for pointing that out. 
* Update the following components to use the newly created GA4 selectors instead of the UA ones. 
    *  `assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidgetGA4/index.js`
    * `assets/js/modules/analytics/components/dashboard/DashboardOverallPageMetricsWidgetGA4.js`
    * Ensure that the `unifiedPagePathScreen` metric is used in case of Entity Dashboard for both widgets. ie. is has `entityURL`. 
* In `assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/Footer.js`
    * Create a `SourceLinkAnalytics4` component that mirrors the logic from `SourceLinkAnalytics` but uses the aforementioned selector and new logic for the updated deep links.
    * Conditionally render `SourceLinkAnalytics` or `SourceLinkAnalytics4` based on whether the current dashboard is GA4 in `Footer` using `isGA4DashboardView` selector from Analytics module.
* In `assets/js/modules/analytics/components/module/ModulePopularPagesWidget/Footer.js`:
    * Use the `getServiceReportURL` from GA4 for the source link when the dashboard is in GA4 view using `isGA4DashboardView` selector from Analytics module. 

### Test Coverage

* Add tests for the newly introduced selectors. 
* Fix any failing tests. 

## QA Brief

- Setup Site Kit and connect Analytics with GA4.
- Ensure the `ga4Reporting` feature flag is enabled, and the Dashboard View is set to Google Analytics 4.
- Check the Source links for each of the widgets specified in the AC navigate to the correct report. Note, there are a couple of aspects of the AC that are not quite correct, as follows:
  - **All Traffic Widget > Locations** (described as **Countries** in the AC):
    - Although the report listed is **Reports > Demographics > Overview > Users by country**, in fact the expected report is **Reports > Demographics > Overview > Demographic details**, with **Country** selected as the detail - i.e. the result of navigating to **Users by country** and clicking on **View countries**. This tallies with the provided URL example in the PRD, and also, it's the report version which is filterable - **Users by country** doesn't allow to filter by the page path.
  - **Search Traffic Widget**:
    - Although only the "Total Users from Search" (i.e. "Unique Visitors from Search") metric is listed in the AC, in fact the same report should be linked to for _all_ of the GA4 metrics in this widget: "Unique Visitors from Search", "Conversions" on the Main Dashboard, and "Engagement Rate" on the Entity Dashboard.
- Verify the currently selected date range is selected in the reports, as well as the described filters.

## Changelog entry

* Update Analytics source links on the GA4 version of the dashboard.

**Comment by marrrmarrr** [comment]
@techanvil I've added a [section in the PRD](https://docs.google.com/document/d/17RKIrdm5PBK8j6rP0GV_2QlztSLXqYQX8ikz3kQmrAo/edit?resourcekey=0-njOU7bn1dhvi5Yp2jeX5zw#bookmark=id.9e5lpgtwjov3
) with a list of links to be updated. 
Let me know if this info is enough, or we need more context.

**Comment by techanvil** [comment]
Thanks @marrrmarrr - I've left a few related questions on the PRD.

**Comment by tofumatt** [comment]
The example links are behind a Google corporate link I can't access, but I think the references are enough; I've gone through and listed example links for all the new metrics in GA4 we can use.

**Comment by techanvil** [comment]
Hey @tofumatt just letting you know, I spotted a reference to `page_url` which was left in the PRD by accident and should in fact be `unifiedPagePathScreen`, so I have updated the AC accordingly (and left a suggestion to update the PRD).

**Comment by techanvil** [comment]
Hey @kuasha420, the IB is looking good. I have a few comments on it, please take a look.

- The AC listed "Top search queries", "Top content" and "Top earning pages" as though they are part of the Search Traffic widget, which is a bit misleading. I should have spotted this earlier, anyhow, I've now corrected the AC, but the IB needs to be updated to include the "Top content" widget.
- It would be useful to provide a bit more technical direction on how to construct the links. It's evident when deconstructing them that the query params contain JSON-encoded objects which are in turn URI-encoded. We don't want to hard code the various encoded values with literal percent signs etc., and might as well specify this in the IB for the sake of clarity.
- The IB reads as though `unifiedPagePathScreen` should only be included on the Entity Dashboard for the `DashboardAllTrafficWidgetGA4` component, but it should be used in the `DashboardOverallPageMetricsWidgetGA4` component too. This is fairly evident when cross-referencing the AC, but it would be best to make it clear in the IB too.

**Comment by kuasha420** [comment]
Thanks @techanvil for the feedback. I have updated the IB to account for all three points you raised. I didn't realize that there was JSON encoded object hiding in the query params! I thought there was only url encoding so didn't feel the need to elaborate further. :disappointed: Thanks for catching it.

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

@techanvil 
Below is the AC for All traffic channels and Locations 
All Traffic Widget:

- If the "Channels" tab is selected, update to [Reports > Life cycle > Acquisition > Traffic](https://support.google.com/analytics/answer/12923437?hl=en&ref_topic=9303476), eg: https://analytics.google.com/analytics/web/#/p${propertyId}/reports/dashboard?params=_u..nav%3Dmaui&r=lifecycle-acquisition-overview&ruid=lifecycle-acquisition-overview,life-cycle,acquisition&collectionId=life-cycle

- If the "Countries" tab is selected, update to [Reports > Demographics > Overview > Users by country](https://support.google.com/analytics/answer/9353692), eg: https://analytics.google.com/analytics/web/#/p${propertyId}/reports/dashboard?params=_u..nav%3Dmaui&r=user-demographics-overview&ruid=user-demographics-overview,user,demographics&collectionId=user`

Both above mentioned URL contain - 'dashboard' whereas actual URL have 'explorer' at the place of dashboard. Example URL `(https://analytics.google.com/analytics/web/?pli=1#/p285794360/reports/explorer?params=_r..nav%3Dmaui%26_u.date00%3D20230308%26_u.date01%3D20230404%26_u.date10%3D20230208%26_u.date11%3D20230307%26_r.explorerCard..selmet%3D%5B%22activeUsers%22%5D%26_r.explorerCard..seldim%3D%5B%22country%22%5D&r=user-demographics-detail&collectionId=user)`

 URL for devices tab contain explorer in AC and it is same as actual URL.
Although, the URL destination is as per AC but want to flag this difference. Is URL mentioned under AC is correct ? and if it correct then should we require changes here in actual URL ?

**Comment by techanvil** [comment]
Hi @mohitwp, thanks for flagging this.

The URLs you've mentioned in the AC are not quite correct. For some reason the URL seems to have been pasted from the Overview section of the relevant report, rather than the details section we actually want to navigate to.

**Reports > Life cycle > Acquisition > Overview**

![image](https://user-images.githubusercontent.com/18395600/230130480-7f7f2329-81e2-438d-8187-f5cbbb1193be.png)

**Reports > Life cycle > Acquisition > Traffic acquisition** (the correct report)

![image](https://user-images.githubusercontent.com/18395600/230130703-50b1ff1b-7bb7-41dd-8a39-987ee126d78f.png)

**Reports > Demographics > Overview**

![image](https://user-images.githubusercontent.com/18395600/230130977-ad308a97-c335-4ebd-97fa-80d0fa5af752.png)

**Reports > Demographics > Demographic details: Country** (the correct report)

![image](https://user-images.githubusercontent.com/18395600/230131371-a7540a3f-02d9-4c0e-b8d7-7a94714f739a.png)

So, in a nutshell the AC is a little off, and the links do navigate to the correct report URLs.
